docs/docs/reference/release-notes/migrating-source-plugin-from-v3-to-v4.md

```diff
@@ -0,0 +1,276 @@
+  if (globalPluginOptions.convert) {
+    return arg.toUpperCase()
+  } else {
+    return arg
+  }
+}
+```
+
+## 4. Custom error map needs to be initiated in onPluginInit
+
+To make it more clear for users when things go wrong, Gatsby has structured logs. It allows plugins to properly guide people to a solution by providing metadata such as urls to docs, … This feature needs to be initialized in onPluginInit to allow GraphQL resolvers to report these errors.
+
+Similar to handling global state, we provide a helper function in `gatsby-plugin-utils` (make sure to add this explicitly to your dependencies) to check if this new API is available. This will help you keep backwards compatibility in V3 while moving forward to a V4 world.We also need to check if we are using the `stable` or `unstable` version of `onPluginInit`.
+
+## The Old Way
+
+```javascript
+const ERROR_MAP = {
+  [CODES.Generic]: {
+    text: context => context.sourceMessage,
+    level: `ERROR`,
+    type: `PLUGIN`,
+  },
+  [CODES.MissingResource]: {
+    text: context => context.sourceMessage,
+    level: `ERROR`,
+    type: `PLUGIN`,
+    category: `USER`,
+  },
+}
+exports.onPreInit = ({ reporter }) => {
+  if (reporter.setErrorMap) {
+    reporter.setErrorMap(ERROR_MAP)
+  }
+}
+```
+
+## The New Way
+
+```javascript
+let coreSupportsOnPluginInit: "unstable" | "stable" | undefined
+
+try {
+  const { isGatsbyNodeLifecycleSupported } = require(`gatsby-plugin-utils`)
+  if (isGatsbyNodeLifecycleSupported(`onPluginInit`)) {
+    coreSupportsOnPluginInit = "stable"
+  } else if (isGatsbyNodeLifecycleSupported(`unstable_onPluginInit`)) {
+    coreSupportsOnPluginInit = "unstable"
+  }
+} catch (e) {
+  console.error(`Could not check if Gatsby supports onPluginInit lifecycle`)
+}
+
+const ERROR_MAP = {
+  // error map
+}
+
+const initializePlugin = ({ reporter }) => {
+  if (reporter.setErrorMap) {
+    reporter.setErrorMap(ERROR_MAP)
+  }
+}
+
+// need to conditionally export otherwise it throws an error for older versions
+if (coreSupportsOnPluginInit === "stable") {
+  exports.onPluginInit = initializePlugin
+} else if (coreSupportsOnPluginInit === "unstable") {
+  exports.unstable_onPluginInit = initializePlugin
+} else {
+  exports.onPreInit = initializePlugin
+}
+```
+
+## Recommendations for publishing your new source plugin version
+
+Publish a new version of their Gatsby 4 compatible package that references gatsby: ^4.0.0-next.0 to signal that the given source plugin version if specifically updated to work with Gatsby 4
+
+- When first publishing, it is recommended that source plugin maintainers indicate their plugin versions are beta ( they publish with the @next tag as well, we can communicate to all our users that v4 for plugins live on the "next" tag )
+- It is okay for source plugin authors to publish their beta versions in advance of Gatsby 4 beta ( perhaps preferred so that we don't have incompatibility as soon as folks attempt to use Gatsby 4 )
```

Comment:
Do we want to include this section at all? Does this make sense to have it here?

Reply:
I only kept the first paragraph, it might be helpful to link to a doc about publishing source plugins (if we have one?)